### PR TITLE
tests(integration) wait for processes to die after killing

### PR DESCRIPTION
**Pull request by hishamhm**
Wait for Kong to actually die after finishing a test. This should considerably reduce the flakiness of integration tests, avoiding results such as the one seen here:

https://travis-ci.org/Mashape/kong/jobs/278614018

This is currently achieved via a sleep-wait using `ps -p`, which is portable enough for POSIX platforms.

Also, remove outdated Serf reference.

### CODE REVIEW

File: spec/helpers.lua [hunk 1]
```diff
@@ -892,6 +892,33 @@ local function wait_for_invalidation(key, timeout)
...
+    local max_time = ngx.now() + (timeout or 10)
+    repeat
+      if pl_utils.execute("ps -p " .. pid .. " >/dev/null 2>&1") then
+        ngx.sleep(0.05)
+      else
+        return
+      end
+    until ngx.now() >= max_time
+    -- Timeout: kill with SIGKILL
+    pl_utils.execute("kill -9 " .. pid .. " >/dev/null 2>&1")
```
Comment (thibaultcha) [inline spec/helpers.lua]: Yes to this addition :+1:

File: spec/helpers.lua [hunk 2]
```diff
@@ -892,6 +892,33 @@ local function wait_for_invalidation(key, timeout)
...
+--- Waits for the termination of a pid.
+-- @param pid_path Filename of the pid file.
+-- @param timeout (optional) in seconds, defaults to 10.
+local function wait_pid(pid_path, timeout)
+  local pid
+  local fd = io.open(pid_path)
+  if fd then
+    pid = fd:read("*l")
+    fd:close()
+  end
```
Comment (thibaultcha) [inline spec/helpers.lua]: Any reason for not using `kong.cmd.utils.kill`? It also has an `is_running()` method available that this function could make use of.

Actually, don't you think this function should live in this `kill` module instead? We could (later) make use of it in `kong stop`/`kong quit` as well.

File: spec/helpers.lua [hunk 3]
```diff
@@ -892,6 +892,33 @@ local function wait_for_invalidation(key, timeout)
...
+
+  if pid then
+    local max_time = ngx.now() + (timeout or 10)
+    repeat
+      if pl_utils.execute("ps -p " .. pid .. " >/dev/null 2>&1") then
+        ngx.sleep(0.05)
+      else
+        return
```
Comment (thibaultcha) [inline spec/helpers.lua]: It is somewhat emphasized in the contributing file that returning early is an appreciated practice. Not a blocker here but I do think it is a good practice to build and enforce:

```lua
if not pl_utils.execute("ps -p " .. pid .. " >/dev/null 2>&1") then
  return
end

-- still running, let's wait some more
ngx.sleep(0.05)
```

File: spec/helpers.lua [hunk 4]
```diff
@@ -892,6 +892,33 @@ local function wait_for_invalidation(key, timeout)
...
+      end
+    until ngx.now() >= max_time
+    -- Timeout: kill with SIGKILL
+    pl_utils.execute("kill -9 " .. pid .. " >/dev/null 2>&1")
+    -- Sanity check: check pid again, but don't loop.
+    wait_pid(pid_path, timeout, false)
```
Comment (thibaultcha) [inline spec/helpers.lua]: It doesn't seem like `wait_pid`'s signature accepts a third argument?

File: spec/helpers.lua [hunk 5]
```diff
@@ -974,12 +1002,12 @@ return {
...
-      if pl_path.exists(pid_path) then
-        kill.kill(pid_path, "-TERM")
-      end
+    -- kill kong_tests.conf service
+    local pid_path = running_conf.nginx_pid
+    if pl_path.exists(pid_path) then
+      kill.kill(pid_path, "-TERM")
+      wait_pid(pid_path, timeout, true)
     end
+
```
Comment (thibaultcha) [inline spec/helpers.lua]: Unrelated change? Curious what the motivation is.

File: spec/helpers.lua (hunk 2 above)
Comment (hishamhm) [reply spec/helpers.lua]: > Any reason for not using kong.cmd.utils.kill? 

It was required locally in the other function that was marked `only use in...`, so I assumed it wasn't desired to fetch this module here. Plus, why load a module do something that could be achieved in one line? I don't need the extra checks and logs of the `kill` module there.

> It also has an is_running() method available that this function could make use of.

Same rationale as the above, plus it uses `kill -0`. `kill` is implemented as a builtin in some shells and I've been too bitten before in incompatibilities in builtins replacing commands, so I figured this would be more reliable.

> Actually, don't you think this function should live in this kill module instead?

Yes, it could!

Comment (thibaultcha) [reply spec/helpers.lua]: > do something that could be achieved in one line?

The `kill` module does all of the pid file reading and command building, which is more than "one line". It also logs to `debug` the attempts it makes and the commands it runs, which could also turn very helpful one day. This is what I meant. Replacing `is_running()`'s `kill -0` with `ps -p` was also on the table when I suggested this, but it definitely wasn't my concern here. Not duplicating the pid file reading logic was - if that makes sense?

File: spec/helpers.lua (hunk 5 above)
Comment (hishamhm) [reply spec/helpers.lua]: which unrelated change, removing `serf_pid`?

File: spec/helpers.lua (hunk 4 above)
Comment (hishamhm) [reply spec/helpers.lua]: Good catch. I toyed with the idea of making `kill -9` optional, but decided it would be an unnecessary moving part.

File: spec/helpers.lua (hunk 2 above)
Comment (hishamhm) [reply spec/helpers.lua]: I explicitly didn't want to rely my loop on a function that reads the pid file. The very thing I'm doing here is making sure that the process is really gone, and I don't want to get a false positive if the pid file has already been deleted but the process is still lingering.

File: spec/helpers.lua (hunk 5 above)
Comment (Tieske) [reply spec/helpers.lua]: new line, whitespace

File: spec/helpers.lua (hunk 2 above)
Comment (thibaultcha) [reply spec/helpers.lua]: @hishamhm Gotcha. Sounds good!

File: spec/helpers.lua (hunk 4 above)
Comment (thibaultcha) [reply spec/helpers.lua]: I think this one would be the only blocker at this point. Other changes can be dealt with. In the future, using this in `kong stop` will also help making the CLI tests (that rely on `kong stop` and `kong start`) more reliable as well.

**Comment by hishamhm** [comment]
Spurious argument dealt with, from my part I'm done with this one.

**Comment by thibaultcha** [comment]
Merged with minor edits. Thanks @hishamhm!
